piksi_tools/ardupilot/mavlink_decode.py

```
@@ -8,92 +8,273 @@
   sudo pip install sbp
 
 """
-from pymavlink.DFReader import DFReader_binary
 from sbp.table import dispatch, _SBP_TABLE
 from sbp.msg import SBP
+from struct import unpack
+from datetime import datetime, timedelta
 
 import json
 
-SBR1_DATASTART = 16
-SBR2_DATASTART = 13
+ARDUPILOT_LOG_HEADER = bytearray([0xA3, 0x95])  # each Ardupilot log frame starts with "0xA3 0x95"
+
+# Timestamp format : 2017-05-26T21:40:15.717000
+TIMESTAMP_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+TIME_ORIGIN = "1980-01-06T00:00:00.000000"
+
+FORMAT_SIZE_BYTES = {'B': 1,
+                     'H': 2,
+                     'I': 4,
+                     'Q': 8,
+                     }
+
+"""
+Format characters in the format string for mavlink binary log messages
+  b   : int8_t
+  B   : uint8_t
+  h   : int16_t
+  H   : uint16_t
+  i   : int32_t
+  I   : uint32_t
+  f   : float
+  d   : double
+  n   : char[4]
+  N   : char[16]
+  Z   : char[64]
+  c   : int16_t * 100
+  C   : uint16_t * 100
+  e   : int32_t * 100
+  E   : uint32_t * 100
+  L   : int32_t latitude/longitude
+  M   : uint8_t flight mode
+  q   : int64_t
+  Q   : uint64_t
+"""
+
+
+def search_binary_key(log, searched_keys):
+  key = bytearray()
+  for k in range(len(searched_keys[0])):
+    key.append(bytes(0x0))
+  while key not in searched_keys:
+      byte = log.read(1)
+      if len(byte):
+        for i in range(len(key)-1):
+          key[i] = key[i+1]
+        key[2] = byte
+      else:
+        return None
+  return key
+
+
+class SBR1:
+
+  HEADER = bytearray([0xA3, 0x95, 0xE5])
+  LABELS = "TimeUS,msg_type,sender_id,msg_len,1,2,3,4,5,6,7,8"
+  FORMAT_HEADER = 'QHHB'  # Data format before binary message
+  FORMAT_BINARY = 'QQQQQQQQ'  # Data format for binary message
+  SIZE = 77
+  BINARY_SIZE = 64
+
+  def __init__(self, binary):
+    self.time_us = None
+    self.msg_type = None
+    self.sender_id = None
+    self.msg_len = None
+    self.msg = None
+    self.read_bytes(binary)
+
+  def read_bytes(self, binary):
```

Lets throw an error or information message if the size is incorrect